File: SiHyeon/week3/키패드 누르기.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

int getGap(int start, int end)
{
    int xGap = abs((start % 3 != 0 ? start / 3 : start / 3 - 1) - end / 3);
    int yGap = abs((start % 3 != 0 ? start % 3 - 1 : 2) - (end % 3 - 1));
    return xGap + yGap;
}

string solution(vector<int> numbers, string hand) {
    //번호판을 2차원 배열로 생각하면 x값끼리의 차 + y값 끼리의 차 = 이동거리
    //x값은 3으로 나눈값 y값은 3으로 나눈 나머지 -1 값이다.
    int left = 10; //*은 위치상 10번 위치
    int right = 12; //#은 위치상 12번 위치
    string answer = "";
    for (int n : numbers)
    {
        n = n == 0 ? 11 : n; //0은 위치상 11번 위치
        switch (n)
        {
        case 1:
        case 4:
        case 7:
            left = n;
            answer += "L";
            break;
        case 3:
        case 6:
        case 9:
            right = n;
            answer += "R";
            break;
        default:
            if (getGap(left, n) == getGap(right, n)) //거리가 같으면
            {
                if (hand == "left")
                {
                    left = n;
                    answer += "L";
                }
                else
                {
                    right = n;
                    answer += "R";
                }
            }
            else if (getGap(left, n) < getGap(right, n))
            {
                left = n;
                answer += "L";
            }
            else
            {
                right = n;
                answer += "R";
            }
            break;
        }
    }
    return answer;
}

File: yejee/week3/키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        String answer = "";
        int left=10, right =12;
        for(int i = 0 ; i < numbers.length ; i++){
            if(numbers[i] == 0)
                numbers[i] = 11;
            if(numbers[i] % 3 == 1){
                answer += "L";
                left = numbers[i];
            }
            else if(numbers[i] % 3 == 0){
                answer += "R";
                right = numbers[i];
            }
            else{
                answer += LorR(left, right, numbers[i], hand);
                if(LorR(left, right, numbers[i], hand)=="L")
                    left = numbers[i];
                else
                    right = numbers[i];
            }
        }        
        return answer;
    }
    public String LorR(int left, int right, int next, String hand){
        int dis_l, dis_r;
        dis_l = (Math.abs(left - next) + 1) / 3;
        dis_r = (Math.abs(right - next) + 1) / 3;
        if((left-next) % 3 != 0)
            dis_l++;
        if((right-next) % 3 != 0)
            dis_r++;
        if(dis_l < dis_r){
            return "L";}
        else if(dis_l > dis_r){
            return "R";}
        else{ //양손의 거리가 같은 경우
            if(hand.equals("right"))
                return "R";
            else
                return "L";
        }
    }
    
}